### PR TITLE
Away users

### PR DESCRIPTION
**Pull request by jackleyden**
This request contains an update to the user selection process.  This feature looks at when a user is going to be away in order to exclude them from task assignment.

Changes:
- Added user state
- Add List#ordered_users.
- Add instigator to tasks.
- Added List#instigator_excluded.
- Remove user mandatory in Task.
- Add Task#assign_user.
- Update CreateTaskOperation to use user assignment.
- Remove List#next_user.

### CODE REVIEW

File: app/operations/create_task_operation.rb [hunk 1]
```diff
@@ -5,11 +5,17 @@ def call(params)
     list = List.find_by!(webhook_token: params[:list_webhook_token])
     task = list.tasks.build(
       description: params[:description],
-      user: list.next_user(params[:instigator_slack_id]),
+      instigator: User.find_by(slack_id: params[:instigator_slack_id])
```
Comment (atcruice) [inline app/operations/create_task_operation.rb]: Missing trailing comma.

File: db/migrate/20180420032328_add_userstate_to_users.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddUserstateToUsers < ActiveRecord::Migration[5.1]
+  def change
+    add_column :users, :excluded_at, :datetime
+  end
+end
```
Comment (atcruice) [inline db/migrate/20180420032328_add_userstate_to_users.rb]: The term _state_ is too general. Programming is all about state. Let's be more specific with the terminology. Amend:
- the name of this migration `AddExcludedAtToUsers`
- the filename of this migration (will need to rollback before making changes)
- the name of this commit (I'll show you how to make an empty fixup commit to remind you of something)

File: spec/fixtures/users.yml [hunk 3]
```diff
@@ -9,7 +9,7 @@ alex:
 dave:
   slack_id: U3
   slack_name: dave
-
+
```
Comment (atcruice) [inline spec/fixtures/users.yml]: Remove the trailing whitespace, leave the blank line.

File: spec/models/list_spec.rb [hunk 4]
```diff
@@ -16,120 +16,76 @@
       .not_to change { list.webhook_token }
   end
 
-  describe "#next_user" do
+
```
Comment (atcruice) [inline spec/models/list_spec.rb:19]: Remove extra blank line.

File: spec/models/list_spec.rb [hunk 5]
```diff
@@ -16,120 +16,76 @@
...
 
-    it "is the first user when no previous users have been assigned" do
-      expect(list.tasks.size).to eq(0)
-      expect(list.next_user).to eq(users(:aldhsu))
+    it "returns alphabetical users when no tasks exist" do
+      expect(list.ordered_users).to eq([users(:aldhsu), users(:alex), users(:dave), users(:tate)])
     end
 
-    it "is the second user when the first user was previously assigned" do
+    it "returns reordered users when 1 task exists" do
```
Comment (atcruice) [inline spec/models/list_spec.rb]: Update example description with something a little more concise.
```
it "returns alphabetical users rotated once when 1 task exists" do
```

File: spec/models/list_spec.rb [hunk 6]
```diff
@@ -16,120 +16,76 @@
...
+      expect(list.ordered_users).to eq([users(:dave), users(:tate), users(:aldhsu), users(:alex)])
     end
 
-    it "is still the second user when the second user archived" do
+    it "returns alphabetical users when all users have tasks" do
+      create_tasks(:aldhsu, :alex, :dave, :tate)
+      expect(list.ordered_users).to eq([users(:aldhsu), users(:alex), users(:dave), users(:tate)])
+    end
+
+    it "returns the second user when they archived last" do
```
Comment (atcruice) [inline spec/models/list_spec.rb]: Update example description with something that communicates better:
```
it "is not affected by archived tasks" do
```

File: spec/models/list_spec.rb [hunk 7]
```diff
@@ -16,120 +16,76 @@
...
       _, second = create_tasks(:aldhsu, :alex)
       second.archived!
       expect(list.tasks.size).to eq(2)
-      expect(list.next_user).to eq(users(:alex))
+      expect(list.ordered_users).to eq([users(:alex), users(:dave), users(:tate), users(:aldhsu)])
     end
 
-    it "is also the second user when the first task has not been actioned" do
+    it "returns the second user when the first task has not been actioned" do
```
Comment (atcruice) [inline spec/models/list_spec.rb]: Similar here:
```
it "is not affected by un-actioned tasks" do
```

File: spec/models/task_spec.rb [hunk 8]
```diff
@@ -16,4 +18,121 @@
...
+        user: users(:alex),
+        instigator: nil,
+        list: lists(:outofdate),
+      )
+      expect(task.save).to eq(true)
+    end
+  end
+
+  describe "user not mandatory" do
+    it "doesn't need users value to be full" do
```
Comment (atcruice) [inline spec/models/task_spec.rb]: Update this to match the pattern above.
```
describe "#user_id" do
  it "may not have a user" do
    …
```

File: spec/models/task_spec.rb [hunk 9]
```diff
@@ -16,4 +18,121 @@
...
+        description: "asdf",
+        list: list,
+        instigator: users(:aldhsu),
+        user: nil,
+      )
+      expect(list.ordered_users.first).to eq(users(:aldhsu))
+      expect(task.assign_user).to eq(users(:aldhsu))
+    end
+
+    it "must eliminate any excluded users and instigator" do
```
Comment (atcruice) [inline spec/models/task_spec.rb]: s/must eliminate/eliminates

File: spec/models/task_spec.rb [hunk 10]
```diff
@@ -16,4 +18,121 @@
...
+      alex = users(:alex)
+      alex.update!(excluded_at: Time.current.beginning_of_minute)
+
+      expect(list.ordered_users.first).to eq(users(:aldhsu))
+      expect(list.ordered_users.second).to eq(users(:alex))
+      expect(alex.excluded?).to eq(true)
+      expect(task.assign_user).to eq(users(:dave))
+    end
+
+    it "must eliminate any excluded users and not the instigator" do
```
Comment (atcruice) [inline spec/models/task_spec.rb]: s/must eliminate/eliminates

File: spec/models/task_spec.rb [hunk 11]
```diff
@@ -16,4 +18,121 @@
...
+    it "returns nil if no users in list" do
+      list = lists(:empty)
+      task = Task.new(
+        description: "asdf",
+        list: list,
+        instigator: users(:alex),
+        user: nil,
+      )
+      expect(task.assign_user).to eq(nil)
+
```
Comment (atcruice) [inline spec/models/task_spec.rb]: Remove blank line.

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    user = users(:aldhsu)
+    user.update!(excluded_at: nil)
+    expect(user.excluded?).to eq(false)
+  end
+
+  it "returns false for future excluded user data" do
+    user = users(:aldhsu)
+    user.update!(excluded_at: Time.current.advance(years: 1))
+    expect(user.excluded?).to eq(false)
+  end
```
Comment (atcruice) [inline spec/models/user_spec.rb]: Missing a containing `describe` block for these examples:
```
RSpec.describe User, type: :model do
  fixtures :users

  describe "#excluded?" do # <<<<<<<
    it "returns true for current exc…
```

**Review by atcruice — CHANGES_REQUESTED** [review]
@Grand-Admiral Minor changes, mostly rewording and simple tidying. No architectural changes.

File: app/operations/create_task_operation.rb (hunk 1 above)
Comment (jackleyden) [reply app/operations/create_task_operation.rb]: changes made at:
45061af13cabb93d0eaec1ab38df15f94198afec

File: db/migrate/20180420032328_add_userstate_to_users.rb (hunk 2 above)
Comment (jackleyden) [reply db/migrate/20180420032328_add_userstate_to_users.rb]: changes at:
cf6f4e5a6d00633cd578276c011cee0436d6debd

File: spec/fixtures/users.yml (hunk 3 above)
Comment (jackleyden) [reply spec/fixtures/users.yml]: Changes at:
192c98653d079a0385dd613224ef2bdfdbdbaded

File: spec/models/list_spec.rb (hunk 4 above)
Comment (jackleyden) [reply spec/models/list_spec.rb:19]: Change at:
2978afb42b654abc6acbb9ace07a2833d6afac3e

File: spec/models/list_spec.rb (hunk 5 above)
Comment (jackleyden) [reply spec/models/list_spec.rb]: Find at:
4fdc1c1b6b4750311aae893618ce59e61b5ab70e

File: spec/models/task_spec.rb (hunk 8 above)
Comment (jackleyden) [reply spec/models/task_spec.rb]: Find at:
77261f210e33d61849c42ffc82a2c996db4a5bed

File: spec/models/task_spec.rb (hunk 9 above)
Comment (jackleyden) [reply spec/models/task_spec.rb]: Find at:
abc8dcd8e83a8dd9b469f70e077cdd02dc5f73d4

File: spec/models/task_spec.rb (hunk 10 above)
Comment (jackleyden) [reply spec/models/task_spec.rb]: Find at:
abc8dcd8e83a8dd9b469f70e077cdd02dc5f73d4

File: spec/models/task_spec.rb (hunk 11 above)
Comment (jackleyden) [reply spec/models/task_spec.rb]: Find at:
abc8dcd8e83a8dd9b469f70e077cdd02dc5f73d4

File: spec/models/user_spec.rb (hunk 12 above)
Comment (jackleyden) [reply spec/models/user_spec.rb]: Find at:
71b8bd384d188e4d0f583f5083471562abc8a167

File: spec/models/list_spec.rb (hunk 6 above)
Comment (atcruice) [reply spec/models/list_spec.rb]: Actioned in 0b3bc8e18e5015f4f12370684ed2673d7dc519f4

File: spec/models/list_spec.rb (hunk 7 above)
Comment (atcruice) [reply spec/models/list_spec.rb]: Actioned in 0b3bc8e18e5015f4f12370684ed2673d7dc519f4

File: spec/models/list_spec.rb (hunk 6 above)
Comment (jackleyden) [reply spec/models/list_spec.rb]: Find at:
0b3bc8e18e5015f4f12370684ed2673d7dc519f4

File: spec/models/list_spec.rb (hunk 7 above)
Comment (jackleyden) [reply spec/models/list_spec.rb]: Find at:
0b3bc8e18e5015f4f12370684ed2673d7dc519f4

**Review by atcruice — APPROVED** [review]
@Grand-Admiral 🚢 You now want to rebase on the latest master, squashing in the fixups, reword that commit message and remove the empty commit. Then force push it up again.

**Pull request merged**